opacus/utils/per_sample_gradients_utils.py

```diff
@@ -0,0 +1,393 @@
+
+    microbatch_grad_samples = {
+        name: p.microbatch_grad_sample
+        for name, p in trainable_parameters(module.wrapped_module)
+    }
+    return microbatch_grad_samples
+
+
+def compute_opacus_grad_sample(
+    x: Union[torch.Tensor, PackedSequence],
+    module: nn.Module,
+    batch_first=True,
+    loss_reduction="mean",
+    grad_sample_mode="hooks",
+) -> Dict[str, torch.tensor]:
+    """
+    Runs Opacus to compute per-sample gradients and return them for testing purposes.
+
+    Args:
+        x: The tensor in input to the ``module``
+        module: The ``ModelWithLoss`` that wraps the nn.Module you want to test.
+        batch_first: Whether batch size is the first dimension (as opposed to the second).
+            Defaults to True.
+        loss_reduction: What reduction to apply to the loss. Defaults to "mean".
+        grad_sample_mode: What sampling method to use to get gradients.
+
+    Returns:
+        Dictionary mapping parameter_name -> per-sample-gradient for that parameter
+    """
+    torch.use_deterministic_algorithms(True)
+    torch.manual_seed(0)
+    np.random.seed(0)
+
+    gs_module = wrap_model(
+        model=clone_module(module),
+        grad_sample_mode=grad_sample_mode,
+        batch_first=batch_first,
+        loss_reduction=loss_reduction,
+    )
+    grad_sample_module = ModelWithLoss(gs_module, loss_reduction)
+
+    grad_sample_module.zero_grad()
+    loss = grad_sample_module(x)
+    loss.backward()
+
+    opacus_grad_samples = {
+        name: p.grad_sample
+        for name, p in trainable_parameters(grad_sample_module.wrapped_module._module)
+    }
+
+    return opacus_grad_samples
+
+
+def check_torch_version_for_ew_sample() -> bool:
+    return torch.__version__ >= (1, 13)
+
+
+def get_grad_sample_modes(use_ew: bool = False):
+    grad_sample_modes = ["hooks", "functorch"]
+    if use_ew and check_torch_version_for_ew_sample():
+        grad_sample_modes.append("ew")
+    return grad_sample_modes
+
+
+def check_per_sample_gradients_are_correct(
+    x: Union[torch.Tensor, PackedSequence],
+    module: nn.Module,
+    *,
+    batch_first=True,
+    atol=10e-6,
+    rtol=10e-5,
+    grad_sample_mode="hooks",
+) -> bool:
+    """
+    A utility to check whether per sample gradients are computed correctly with a particular model.
+    Args:
+        x: The tensor in input to the ``module``
+        module: The ``ModelWithLoss`` that wraps the nn.Module you want to check.
+        batch_first: Whether batch size is the first dimension (as opposed to the second).
+            Defaults to True.
+        atol: The relative tolerance parameter (numpy).
+        rtol: The absolute tolerance parameter (numpy).
+        grad_sample_mode: What sampling method to use to get gradients.
+
+    Returns: True if per sample gradients were computed correctly. False otherwise.
+    """
+    if grad_sample_mode == "functorch":
+        import functorch  # noqa
```

Comment:
I'll move it to the global imports since `functorch` is in `requirements.txt`.

Reply:
That's not a good idea, because of a known functorch bug. 
We import it at the point of usage (opacus/grad_sample/functorch.py) so that people can still use opacus if they don't actively use functorch and their PyTorch version is incompatible

Reply:
What I meant by this question is - why do you need to import it here? You don't seem to be using it